### PR TITLE
Change behaviour to reshape on setValue and observe

### PR DESCRIPTION
**Pull request by gordoncaleb**
Previously, when you observed/setValue/setAndCascade a primitive or array then it would NOT reshape the vertex. This has been confusing to users (and me) and I propose we change the behaviour to be such that anything you set or observe is exactly reflected in the vertex. For example, setting or observing a scalar value would make the shape of the vertex scalar and if you observe an array then the vertex would would be reshaped as a vector.

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java [hunk 1]
```diff
@@ -66,4 +66,35 @@ public void doesLambda() {
...
+        int[] values = new int[]{1, 2, 3};
+        binomialVertex.setAndCascade(values);
+        assertArrayEquals(values, binomialVertex.getValue().asFlatIntegerArray());
+    }
+
+    @Test
+    public void canSetValueAsArrayOfValues() {
+        IntegerVertex binomialVertex = new BinomialVertex(0.5, 20);
+        int[] values = new int[]{1, 2, 3};
+        binomialVertex.setAndCascade(values);
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java]: copy paste error? should be .set()?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java [hunk 2]
```diff
@@ -66,4 +66,35 @@ public void doesLambda() {
...
+    }
+
+    @Test
+    public void canSetValueAsScalarOnNonScalarVertex() {
+        IntegerVertex binomialVertex = new BinomialVertex(new int[]{2, 1}, 0.5, 20);
+        binomialVertex.setValue(2);
+        assertArrayEquals(new int[]{2}, binomialVertex.getValue().asFlatIntegerArray());
+    }
+
 }
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java]: add a `canSetAndCascadeAsScalarOnNonScalarVertex` method (as in the `DoubleVertexTest`)?

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java [hunk 3]
```diff
@@ -46,9 +47,9 @@ public SimpleBooleanTensor(int[] shape) {
 
     public SimpleBooleanTensor(boolean constant, int[] shape) {
         this.data = new boolean[(int) TensorShape.getLength(shape)];
+        Arrays.fill(this.data, constant);
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java]: What's the difference between `System.arraycopy` and `Arrays.fill`? Why are you using both in this class?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/bool/BoolVertexTest.java [hunk 4]
```diff
@@ -95,6 +94,36 @@ public void constantVertexWorksAsExpected() {
...
+        flip.setValue(values);
+        assertArrayEquals(new Boolean[]{true, false, true}, flip.getValue().asFlatArray());
+    }
+
+    @Test
+    public void canSetValueAsScalarOnNonScalarVertex() {
+        BoolVertex flip = new Flip(new int[]{2, 1}, 0.5);
+        flip.setValue(true);
+        assertArrayEquals(new Boolean[]{true}, flip.getValue().asFlatArray());
+    }
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/bool/BoolVertexTest.java]: add a canSetAndCascadeAsScalarOnNonScalarVertex method (as in the DoubleVertexTest)?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java (hunk 1 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java]: Good catch.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java (hunk 2 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/IntegerVertexTest.java]: Yup. Good call.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java (hunk 3 above)
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java]: Fill is for filling a single value into a data buffer for an arbitrary shape e.g. a 2x2 of trues. Array copy is used to copy from the user provided boolean buffer to an internal buffer. This copy protects against the case where you pass in more data than the shape allows.

**Pull request merged**